Clean the LLVM wrapper

dpaoliello opened the pull request:
This change sets the visibility for LLVM wrapper FFI items to `pub(crate)`, which allows Rust's dead code lints to flag unused parts of the LLVM wrapper.

dpaoliello committed:
Clean llvm wrapper

dpaoliello commented on compiler/rustc_codegen_llvm/src/llvm/enzyme_ffi.rs:
```diff
@@ -1,28 +1,28 @@
-#![allow(non_camel_case_types)]
+#![allow(non_camel_case_types, dead_code)]
```
Note to reviewers: seems like this is new code, but a bunch of it is dead. Should I also clean this file, or are there plans to expand uses in the future?

workingjubilee replied:
Please leave it in place for now as @ZuseZ4 is working on completing this feature and it has been split over multiple PRs due to being a huge lift.

klensy commented:
cc @Zalathar #134009

Zalathar commented on compiler/rustc_codegen_llvm/src/llvm/ffi.rs:
```diff
@@ -170,16 +155,15 @@ impl Visibility {
 
 /// LLVMUnnamedAddr
 #[repr(C)]
-pub enum UnnamedAddr {
-    No,
-    Local,
-    Global,
```
For enums that are passed across FFI to C/C++ code, we shouldn't delete Rust-side variants that are unused.

Doing so just makes it harder to verify that the Rust-side enum and the C-side enum have the same layout, which is essential for soundness.

Zalathar commented on compiler/rustc_codegen_llvm/src/llvm/ffi.rs:
```diff
@@ -819,9 +775,8 @@ pub mod debuginfo {
-    pub enum DebugNameTableKind {
+    pub(crate) enum DebugNameTableKind {
         Default,
-        Gnu,
```
This is another case where removing unused enum variants is a net negative. It just makes auditing harder.

dpaoliello replied:
In this case, it was removed from both the Rust and C++ side: would you rather that I add it back to both?

Zalathar replied:
Yeah, please add it back to both.

(Our `LLVMRustDebugNameTableKind` enum is basically just a substitute for LLVM's `DICompileUnit::DebugNameTableKind`. It's needed to work around the fact that we can't reliably pass LLVM's C++ enums across FFI, because the LLVM C++ API isn't stable, so upgrading LLVM would potentially introduce silent UB if the enums change incompatibly.)

dpaoliello replied:
Done

Zalathar commented:
(There's more review I want to do here, but unfortunately GitHub has decided that trying to add a review comment should now hang the browser for several seconds, so the review interface is more-or-less unusable right now.)

klensy commented:
I've tried some of this cleanup in #127274, so some review comments from it can be used here as well.

dpaoliello commented:
> I've tried some of this cleanup in #127274, so some review comments from it can be used here as well.

Good call:
* Switched to `#[expect(dead_code)]` where possible.
* I've added back in the enum variants that I removed, so #85677 doesn't apply.
* I think removing the tail-call stuff is still reasonable. It was added in Jul 2023 via #112791 but is still not being used for anything. It would be trivial to add it back if required.

workingjubilee commented:
Yes, removing the tail call stuff is fine, as you said it is a trivial revert.

lcnr commented:
r? @Zalathar as you're already reviewing this PR :thinking: feel free to reassign

Zalathar commented:
I'm in favour of removing unnecessary `pub`, but I think this PR tries to do too much all at once, which makes it hard to review.

I would suggest trying to split it up into smaller self-contained PRs, e.g.:
- A PR to remove things that are clearly unused (e.g. `LLVMMachineType`, `LLVMRustTailCallKind`, `LLVMRustIsBitcode`), without any other changes.
- Individual PRs that incrementally remove pub from smaller chunks of the LLVM or LLVMRust bindings, without trying to change everything at once.

(I would also be perfectly happy to write some of these PRs myself, but if you are keen to write them then I can review them myself without having to rely on another reviewer, so either approach is fine with me.)

dpaoliello commented:
> I'm in favour of removing unnecessary `pub`, but I think this PR tries to do too much all at once, which makes it hard to review.
> 
> I would suggest trying to split it up into smaller self-contained PRs, e.g.:
> 
> * A PR to remove things that are clearly unused (e.g. `LLVMMachineType`, `LLVMRustTailCallKind`, `LLVMRustIsBitcode`), without any other changes.
> * Individual PRs that incrementally remove pub from smaller chunks of the LLVM or LLVMRust bindings, without trying to change everything at once.
> 
> (I would also be perfectly happy to write some of these PRs myself, but if you are keen to write them then I can review them myself without having to rely on another reviewer, so either approach is fine with me.)

Fair enough: I've converted this to a draft and will send out sliced up PRs as I have time

dpaoliello commented:
First chunk, remove the dead code: #136653

bors commented:
:umbrella: The latest upstream changes (presumably #136697) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
